### PR TITLE
Refactor FastXYBToSRGB8 to use an intermediate int16 buffer.

### PR DESCRIPTION
**Pull request by veluca93**
This is in preparation of running more computation in 16-bit fixpoint.

Slight slowdown on fd4 (22.2MP/s -> 21.7MP/s).

### CODE REVIEW

File: lib/jxl/dec_xyb-inl.h [hunk 1]
```diff
@@ -86,10 +86,34 @@ static inline HWY_MAYBE_UNUSED bool HasFastXYBTosRGB8() {
 #endif
 }
 
-static inline HWY_MAYBE_UNUSED void FastXYBTosRGB8(
-    const Image3F& input, const Rect& input_rect, const Rect& output_buf_rect,
-    const ImageF* alpha, const Rect& alpha_rect, bool is_rgba,
-    uint8_t* JXL_RESTRICT output_buf, size_t xsize, size_t output_stride) {
+template <size_t shift>
```
Comment (deymo) [inline lib/jxl/dec_xyb-inl.h:89]: Would be good to explain the requirements of this function; such as `in` needs xsize rounded up to 4 floats; and `out`needs round up to 8 int16_t, right?

File: lib/jxl/dec_xyb-inl.h [hunk 2]
```diff
@@ -86,10 +86,34 @@ static inline HWY_MAYBE_UNUSED bool HasFastXYBTosRGB8() {
...
+static inline HWY_MAYBE_UNUSED void ConvertToFixpoint(const float* in,
+                                                      int16_t* out,
+                                                      size_t xsize) {
+#if HWY_TARGET == HWY_NEON
+  for (size_t x = 0; x < xsize; x += 8) {
+    // TODO(veluca): consider different approaches here, like vld1q_f32_x2.
+    float32x4_t flt_left = vld1q_f32(in + x);
+    int16x4_t int16_left = vqmovn_s32(vcvtq_n_s32_f32(flt_left, shift));
+    float32x4_t flt_right = vld1q_f32(in + x + (x + 4 < xsize ? 4 : 0));
```
Comment (deymo) [inline lib/jxl/dec_xyb-inl.h:98]: 🤯

**Review by deymo — APPROVED** [review]

File: lib/jxl/dec_xyb-inl.h (hunk 1 above)
Comment (veluca93) [reply lib/jxl/dec_xyb-inl.h:89]: No need for that for out - no more than xsize outputs are written.

File: lib/jxl/dec_cache.h [hunk 3]
```diff
@@ -205,6 +209,12 @@ struct PassesDecoderState {
                                       kGroupDim + 2 * kGroupDataYBorder);
       }
     }
+    if (fast_xyb_srgb8_conversion) {
+      for (size_t _ = fixpoint_srgb_buffer.size(); _ < num_threads; _++) {
```
Comment (sboukortt) [inline lib/jxl/dec_cache.h:213]: Would it make sense to `reserve`?

**Review by sboukortt — APPROVED** [review]

File: lib/jxl/dec_cache.h (hunk 3 above)
Comment (veluca93) [reply lib/jxl/dec_cache.h:213]: I'd say probably not -- but I also think I should actually wait to refactor things a bit before I submit this....

**Pull request closed without merging**